include/pybind11/pybind11.h

```diff
@@ -1501,18 +1501,34 @@ class class_ : public detail::generic_type {
     }
 
 private:
+    template <typename T>
```

Comment:
While we're at it, what do you think about moving the code out into pybind11::detail (best IMO), or at least mark it as public. There is a chance I may need it for the `smart_holder` code, too. I don't see a reason for hiding/protecting it in an any way.

Reply:
Oh, I forgot to write down one more suggestion: I'd name this `try_get_shared_from_this` or `get_shared_from_this_if_possible`, `optional_get_shared_from_this`, `get_shared_from_this_or_nullptr`, ...

Reply:
Hmmm, public, I don't know, indeed, since this is not an API we really want to offer, as pybind11? `detail` makes sense, if you think you can use it in #2672. (Normally, I'd argue refactoring should be part of the PR that needs it, but this is a bit of a special case, probably ;-) )

> Oh, I forgot to write down one more suggestion: I'd name this `try_get_shared_from_this` or `get_shared_from_this_if_possible`, `optional_get_shared_from_this`, `get_shared_from_this_or_nullptr`, ...

Yes! When it's public, out of the scope of this class, definitely! I was first thinking of `safely_get_shared_from_this`, when refactoring, but that's not really true (as it's not officially "safe" pre-C++17).

`maybe_get_shared_from_this` or `try_get_shared_from_this` sounds good to me, yes.